### PR TITLE
CP-23986: Implement QMP event thread in xenopsd

### PR DESCRIPTION
**Pull request by fillzero**
Signed-off-by: Liang Dai <liang.dai1@citrix.com>

### CODE REVIEW

**Comment by fillzero** [comment]
This PR depends on https://github.com/xapi-project/xenopsd/pull/376

**Comment by lindig** [comment]
The design is currently undergoing a lot of changes. I'm not sure what to do with this PR. Maybe it makes sense to do this for prototyping but this code will have to go into a different place.

File: xc/device.ml [hunk 1]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+      debug "Get QMP event, domain-%d: %s" domid qmp_event.event
+
+    let qmp_event_thread () =
+      Monitor.add m pipe_r;
+      while true do
+      try
+        let cs = Hashtbl.fold (fun k v acc -> v :: acc) Lookup.dtof [] in
+        let fds = List.map (fun x -> Qmp_protocol.to_fd x, x) cs in
+        let rs = Monitor.wait m in
+        rs |> List.iter (
```
Comment (mg12) [inline xc/device.ml]: If we only use rs here, then I think it's better to use
```Monitor.wait m |> List.iter (```

File: xc/device.ml [hunk 2]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+    let channel_r = Unix.in_channel_of_descr pipe_r
+    let channel_w = Unix.out_channel_of_descr pipe_w
+
+    let epfd = Epoll.create1 [Epoll.EPOLL_CLOEXEC]
+
+    module Lookup = struct
+      let ftod, dtof = Hashtbl.create 16, Hashtbl.create 16
+      let add fd domid =
+        Hashtbl.replace ftod fd domid;
+        Hashtbl.replace dtof domid fd
```
Comment (mg12) [inline xc/device.ml]: let add c domid =
    Hashtbl.replace ftod (Qmp_protocol.to_fd c) domid;
    Hashtbl.replace dtoc domid c

File: xc/device.ml [hunk 3]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+      let remove fd domid =
+        Hashtbl.remove ftod fd;
+        Hashtbl.remove dtof domid
```
Comment (mg12) [inline xc/device.ml]: let remove c domid =
  Hashtbl.remove ftod (Qmp_protocol.to_fd c);
  Hashtbl.remove dtoc domid

File: xc/device.ml [hunk 4]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+      let domid_of fd = try Some (Hashtbl.find ftod fd) with Not_found -> None
+      let fd_of domid = try Some (Hashtbl.find dtof domid) with Not_found -> None
```
Comment (mg12) [inline xc/device.ml]: let channel_of domid = try Some (Hashtbl.find dtoc domid) with Not_found -> None

File: xc/device.ml [hunk 5]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+      let with_event ~ok ~err = function | Epoll.EPOLLIN -> ok() | e -> err e
+    end
+    let m = Monitor.create ()
+
+    let monitor_path domid = Printf.sprintf "/var/run/xen/qmp-event-%d" domid
+    let debug_exn msg e = debug "%s: %s" msg (Printexc.to_string e)
+
+    let remove domid =
+      try
+        match Lookup.fd_of domid with
```
Comment (mg12) [inline xc/device.ml]: match Lookup.channel_of domid with

File: xc/device.ml [hunk 6]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+        | None -> ()
+        | Some c -> Lookup.remove c domid;
+          Monitor.remove m (Qmp_protocol.to_fd c);
+          Qmp_protocol.close c;
+          debug "Removed QMP Event fd for domain %d" domid
+      with e -> debug_exn (Printf.sprintf "Get exception trying remove qmp on domain-%d" domid) e
+
+    let add domid =
+      try
+        match Lookup.fd_of domid with
```
Comment (mg12) [inline xc/device.ml]: match Lookup.channel_of domid with

File: xc/device.ml [hunk 7]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+                ~ok:(fun () ->
+                       let domid = input_line channel_r in
+                       try add (int_of_string domid)
+                       with e -> debug_exn (Printf.sprintf "Cannot add domain-%s to qmp_event_thread" domid) e)
+                ~err:(fun _ ->
+                        debug "Received unexpected epoll event on pipe_r in qmp_event_thread")
+              )
+            else
+              let c = List.assoc fd fds in
+              match Lookup.domid_of c with
```
Comment (mg12) [inline xc/device.ml]: ~let c = List.assoc fd fds in~
match Lookup.domid_of fd with

File: xc/device.ml [hunk 8]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+        rs |> List.iter (
+          fun (events, fd) ->
+            if fd = pipe_r then
+              events |> List.iter (Monitor.with_event
+                ~ok:(fun () ->
+                       let domid = input_line channel_r in
+                       try add (int_of_string domid)
+                       with e -> debug_exn (Printf.sprintf "Cannot add domain-%s to qmp_event_thread" domid) e)
+                ~err:(fun _ ->
+                        debug "Received unexpected epoll event on pipe_r in qmp_event_thread")
```
Comment (mg12) [inline xc/device.ml]: ~err:(fun e ->
            debug_exn "Received unexpected epoll event on pipe_r in qmp_event_thread" e)

File: xc/device.ml [hunk 9]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+                  ~ok:(fun () ->
+                         if Hashtbl.mem Lookup.dtof domid then try
+                         match Qmp_protocol.read c with
+                           | Event e -> qmp_event_handle domid e
+                           | msg -> debug "Got non-event message, domain-%d: %s" domid (string_of_message msg)
+                         with End_of_file ->
+                           debug "domain-%d: end of file, close qmp socket" domid;
+                           remove domid)
+                  ~err:(fun _ -> debug "EPOLL error on domain-%d, close qmp socket" domid;
+                           remove domid)
```
Comment (mg12) [inline xc/device.ml]: ~err:(fun e -> debug_exn (Printf.sprintf "EPOLL error on domain-%d, close qmp socket" domid) e;
    remove domid)

File: xc/device.ml [hunk 10]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+                  )
+        )
+      with e ->
+        debug "Exception in qmp_event_thread: %s" (Printexc.to_string e);
```
Comment (mg12) [inline xc/device.ml]: debug_exn "Exception in qmp_event_thread" e

File: xc/device.ml [hunk 11]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+        let fds = List.map (fun x -> Qmp_protocol.to_fd x, x) cs in
+        let rs = Monitor.wait m in
+        rs |> List.iter (
+          fun (events, fd) ->
+            if fd = pipe_r then
+              events |> List.iter (Monitor.with_event
+                ~ok:(fun () ->
+                       let domid = input_line channel_r in
+                       try add (int_of_string domid)
+                       with e -> debug_exn (Printf.sprintf "Cannot add domain-%s to qmp_event_thread" domid) e)
```
Comment (mg12) [inline xc/device.ml]: add() is already handled in the wakeup() function.
~let domid = input_line channel_r in~
~try add (int_of_string domid)~
~with e -> debug_exn (Printf.sprintf "Cannot add domain-%s to qmp_event_thread" domid) e)~

File: xc/device.ml [hunk 12]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+          Lookup.add c domid;
+          Monitor.add m (Qmp_protocol.to_fd c);
+          debug "Added QMP Event fd for domain %d" domid
+      with e ->
+        debug_exn (Printf.sprintf "QMP domain-%d: negotiation failed: removing socket" domid) e;
+        remove domid
+
+    let wakeup domid =
+        output_string channel_w (Printf.sprintf "%d\n" domid);
+        flush channel_w
```
Comment (mg12) [inline xc/device.ml]: ~output_string channel_w (Printf.sprintf "%d\n" domid);~
~flush channel_w~
add domid;
(* write single byte to wake up Monitor.wait *)
Unix.write pipe_w " " 0 1

File: xc/device.ml [hunk 13]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
 
 module Dm = struct
 
+  module QMP_Event = struct
+    let (pipe_r, pipe_w) = Unix.pipe ()
+    let channel_r = Unix.in_channel_of_descr pipe_r
+    let channel_w = Unix.out_channel_of_descr pipe_w
```
Comment (mg12) [inline xc/device.ml]: unnecessary (see below)
~let channel_r = Unix.in_channel_of_descr pipe_r~
~let channel_w = Unix.out_channel_of_descr pipe_w~

File: xc/device.ml [hunk 14]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+                ~err:(fun _ ->
+                        debug "Received unexpected epoll event on pipe_r in qmp_event_thread")
+              )
+            else
+              let c = List.assoc fd fds in
+              match Lookup.domid_of c with
+              | None -> ()
+              | Some domid -> events |> List.iter (Monitor.with_event
+                  ~ok:(fun () ->
+                         if Hashtbl.mem Lookup.dtof domid then try
```
Comment (mg12) [inline xc/device.ml]: Lookup.domid_of fd is already checking for that
~if Hashtbl.mem Lookup.dtof domid then try~

File: xc/device.ml [hunk 15]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+                         match Qmp_protocol.read c with
```
Comment (mg12) [inline xc/device.ml]: match Qmp_protocol.read (Lookup.channel_of domid) with

File: xc/device.ml [hunk 16]
```diff
@@ -1430,6 +1431,111 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
...
+                           | Event e -> qmp_event_handle domid e
+                           | msg -> debug "Got non-event message, domain-%d: %s" domid (string_of_message msg)
+                         with End_of_file ->
+                           debug "domain-%d: end of file, close qmp socket" domid;
+                           remove domid)
+                  ~err:(fun _ -> debug "EPOLL error on domain-%d, close qmp socket" domid;
```
Comment (mg12) [inline xc/device.ml]: ~err:(fun e -> debug_exn (Printf.sprintf "EPOLL error on domain-%d, close qmp socket" domid) e;

File: xc/device.ml (hunk 8 above)
Comment (fillzero) [reply xc/device.ml]: Thanks @mg12 , `e` is `Epoll.event` here ...

**Comment by fillzero** [comment]
Cool, thanks @mg12

**Comment by fillzero** [comment]
Closed in favour of https://github.com/xapi-project/xenopsd/pull/378

**Pull request closed without merging**